futures-io/src/lib.rs

```diff
@@ -0,0 +1,305 @@
+//! Asynchronous IO
+//!
+//! This crate contains the `AsyncRead` and `AsyncWrite` traits which allow
+//! data to be read and written asynchronously.
+
+#![no_std]
+#![deny(missing_docs, missing_debug_implementations)]
+#![doc(html_root_url = "https://docs.rs/futures-io/0.2")]
+
+macro_rules! if_std {
+    ($($i:item)*) => ($(
+        #[cfg(feature = "std")]
+        $i
+    )*)
+}
+
+if_std! {
+    extern crate futures_core;
+    extern crate iovec;
+    extern crate std;
+
+    use futures_core::{Async, Poll, task};
+    use std::boxed::Box;
+    use std::io as StdIo;
+    use std::ptr;
+    use std::vec::Vec;
+
+    // Re-export IoVec for convenience
+    pub use iovec::IoVec;
+
+    // Re-export io::Error so that users don't have to deal
+    // with conflicts when `use`ing `futures::io` and `std::io`.
+    pub use StdIo::Error as Error;
+
+    /// A type used to conditionally initialize buffers passed to `AsyncRead`
+    /// methods.
+    #[derive(Debug)]
+    pub struct Initializer(bool);
```

Comment:
Looking at the [std Read::initializer tracking issue](https://github.com/rust-lang/rust/issues/42788#issuecomment-325837435), it seems there is doubt that this is the correct API. In std, the `unstable` attribute protects them.

It seems odd to be using that API here if it's thought to be inferior, especially since we can't rely on `unstable` to protect against breaking changes.

Reply:
It's true that some people have doubts about the API. I personally think that it's the best-designed interface for this that I've seen, and it's the only one I've seen that is sound in the presence of https://github.com/rust-lang/rfcs/pull/2316.